File: 06resources/solution.py
from py


 class Graph:
    def __init__(self):
        self.nodes = {}


    def add_edge(self, from_vertex, to_vertex):
        if from_vertex not in self.nodes:
            self.nodes[from_vertex] = []
        if to_vertex not in self.nodes:
            self.nodes[to_vertex] = OrderedSet()

        self.nodes[from_vertex].append(to_vertex)w



# class Graph {
#     constructor() {
#         this._nodes = new Map();
#     }

#     addEdge(from, to) {
#         if (!this._nodes.has(from)) {
#             this._nodes.set(from, []);
#         }

#         if (!this._nodes.has(to)) {
#             this._nodes.set(to, []);
#         }

#         this._nodes.get(from).push(to);
#     }

#     sortNodes() {
#         for (const [vertex, nodes] of this._nodes) {
#             nodes.sort();
#         }
#     }

#     *[Symbol.iterator]() {
#         const dependencies = this._loadDependencies();
#         for (const dep of dependencies) {
#             yield dep;
#         }
#     }

#     _getStartNodes(used) {
#         let startNodes = new Set(this._nodes.keys());
#         [...this._nodes].forEach(([vertex, ends]) => {
#             if (used.has(vertex)) {
#                 return;
#             }

#             ends.forEach((end) => startNodes.delete(end));
#         });

#         startNodes = [...startNodes].filter((node) => !used.has(node));

#         return startNodes;
#     }

#     _loadDependencies() {
#         const used = new Set();

#         const path = [];

#         while (true) {
#             const startNodes = this._getStartNodes(used);
#             if (startNodes.length === 0) {
#                 break;
#             }

#             const nextNode = min(...startNodes);
#             path.push(nextNode);
#             used.add(nextNode);
#         }

#         return path;
#     }
# }

# const readGraph = () => {
#     const graph = new Graph();
#     const edgesCount = +gets();
#     for (let i = 0; i < edgesCount; i += 1) {
#         const [x, y] = gets().split('needs').map((v) => v.trim());
#         graph.addEdge(x, y);
#     }

#     graph.sortNodes();

#     return graph;
# };

# const solve = () => {
#     const graph = readGraph();
#     return [...graph];
# };

# print(solve().join(" "));

# quit(0);

File: 06resources/solution.cpp
#include<iostream>
#include<vector>
#include<set>
#include<map>

using namespace std;

class Graph {
    public:
        void addEdge(string from, string to);
        vector<string> getSorted();
    private:
        map<string, set< string > > nodes;

        set<string> getStartNodes(set<string> used) {
            set<string> startNodes;
            for(auto vertex = nodes.begin(); vertex != nodes.end(); ++vertex) {
                if (used.find(vertex->first) == used.end()) {
                    startNodes.insert(vertex->first);
                }
            }

            for(auto it = nodes.begin(); it != nodes.end(); ++it) {
                auto key = it->first;
                if(used.find(key) != used.end()) {
                    continue;
                }

                auto neighbors = it->second;

                for(auto next = neighbors.begin(); next != neighbors.end(); ++ next) {
                    startNodes.erase(*next);
                }
            }

            return startNodes;
        }
};

void Graph::addEdge(string from, string to) {
    if(nodes.find(from) == nodes.end()) {
        nodes[from] = set<string>();
    }
    if(nodes.find(to) == nodes.end()) {
        nodes[to] = set<string>();
    }
    nodes[from].insert(to);
}

vector<string> Graph::getSorted() {
    set<string> used;

    vector<string> path;

    while(true) {
        auto startNodes = getStartNodes(used);
        if(startNodes.empty()) {
            break;
        }

        auto vertex = startNodes.begin();

        path.push_back(*vertex);
        used.insert(*vertex);
    }

    return path;
}

Graph readGraph() {
    int m;

    cin >> m;

    Graph graph;

    for(auto i = 0; i < m; ++i) {
        string from, to;
        cin >> from >> to >> to;
        graph.addEdge(from, to);
    }
    return graph;
}

int main() {
    auto graph = readGraph();

    auto result = graph.getSorted();
    for(auto it = result.begin(); it != result.end(); ++ it) {
        cout << *it;
        if(it != result.end() - 1) {
            cout << " ";
        }
    }
    cout << '\n';

    return 0;
}
